src/drivers/intel/ssp/ssp.c

```diff
@@ -273,7 +273,7 @@ static int ssp_set_config_tplg(struct dai *dai, struct ipc_config_dai *common_co
 	if (ssp->state[DAI_DIR_PLAYBACK] > COMP_STATE_READY ||
 	    ssp->state[DAI_DIR_CAPTURE] > COMP_STATE_READY) {
```

Comment:
maybe change this to "== COMP_STATE_ACTIVE"? in case we want to support multiple times of re-configuration before the SSP is started or during 'PAUSED', or reset configuration after 'stopped'.

Reply:
Ranjani changed it from '==' to '>' to avoid some ssp issue.

commit fc73578be774001ef0e9aafdfbc1b4f054cd93d4
Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
Date:   Wed Jul 7 15:56:20 2021 -0700

    drivers: Intel: SSP: ignore config when SSP is already configured

    When two streams are started one after the other, 2 DAI_CONFIG
    IPC's will be sent before the START trigger. This ends up
    configuring the SSP when the first one has already just
    configured it and ends up with xruns.

    The root cause is that the ssp_set_config() function configures
    a set of variables, writes those variables into SSP registers,
    and later on does a read-modify-write operation on the TSRE,
    RSRE and SSE bits.

    If the ssp_set_config is executed more than once, this will
    temporarily clear all three bitfields, and temporarily disable
    DMA transfers and SSP clocks. Avoid this by checking if the
    current state is > COMP_STATE_READY before configuring the
    SSP, so that the ssp_set_config() function only modifies SSP
    registers once.

Reply:
Looks like this logic need look at since the alsa-bat failure is still present.  @brentlu may need some extra logic e.g.
```c
clk:
switch (config->flags & SOF_DAI_CONFIG_FLAGS_MASK) {
	case SOF_DAI_CONFIG_FLAGS_HW_PARAMS:
		if (ssp->params.clks_control & SOF_DAI_INTEL_SSP_CLKCTRL_MCLK_ES) {
			ret = ssp_mclk_prepare_enable(dai);
			if (ret < 0)
				goto out;

			ssp->clk_active |= SSP_CLK_MCLK_ES_REQ;  <<<<<<<<<<<< We could probably check this first and not set M/N if this state is already true.

			dai_info(dai, "ssp_set_config(): hw_params stage: enabled MCLK clocks for SSP%d...",
				 dai->index);
		}

		if (ssp->params.clks_control & SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES) {
			bool enable_sse = false;

			if (!(ssp->clk_active & SSP_CLK_BCLK_ACTIVE))
				enable_sse = true;

			ret = ssp_bclk_prepare_enable(dai);
			if (ret < 0)
				goto out;

			ssp->clk_active |= SSP_CLK_BCLK_ES_REQ;  <<<<<<<<<<<<< ditto as above

			if (enable_sse) {

				/* enable TRSE/RSRE before SSE */
				ssp_update_bits(dai, SSCR1,
						SSCR1_TSRE | SSCR1_RSRE,
						SSCR1_TSRE | SSCR1_RSRE);

				/* enable port */
				ssp_update_bits(dai, SSCR0, SSCR0_SSE, SSCR0_SSE);

				dai_info(dai, "ssp_set_config(): SSE set for SSP%d", dai->index);
			}

			dai_info(dai, "ssp_set_config(): hw_params stage: enabled BCLK clocks for SSP%d...",
				 dai->index);
		}
		break;
```